aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/RetryHandler.java

```diff
@@ -0,0 +1,104 @@
+
+import com.amplifyframework.datastore.utils.ErrorInspector;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.core.SingleEmitter;
+
+/**
+ * Class for retrying call on failure on a single.
+ */
+public class RetryHandler {
+    private final int maxExponentValue = 8;
+    private final int jitterFactorValue = 100;
+    private final int maxAttemptsValue = 3;
+    private final int maxExponent;
+    private final int jitterFactor;
+    private final int maxAttempts;
+
+    /**
+     * Constructor to inject constants for unit testing.
+     * @param maxExponent maxExponent backoff can go to.
+     * @param jitterFactor jitterFactor for backoff.
+     * @param maxAttempts max attempt for retrying.
+     */
+    public RetryHandler(int maxExponent,
+                        int jitterFactor,
+                        int maxAttempts) {
+
+        this.maxExponent = maxExponent;
+        this.jitterFactor = jitterFactor;
+        this.maxAttempts = maxAttempts;
+    }
+
+    /**
+     * Parameter less constructor.
+     */
+    public RetryHandler() {
+        maxExponent = maxExponentValue;
+        jitterFactor = jitterFactorValue;
+        maxAttempts = maxAttemptsValue;
+    }
+
+    /**
+     * retry.
+     * @param single single to be retried.
+     * @param skipExceptions exceptions which should not be retried.
+     * @param <T> The type for single.
+     * @return single of type T.
+     */
+    public <T> Single<T> retry(Single<T> single, List<Class<? extends Throwable>> skipExceptions) {
+        return Single.create(emitter -> call(single, emitter, 0L, maxAttempts, skipExceptions));
+
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private <T> void call(
+            Single<T> single,
+            SingleEmitter<T> emitter,
+            Long delayInSeconds,
+            int attemptsLeft,
+            List<Class<? extends Throwable>> skipExceptions) {
+        single.delaySubscription(delayInSeconds, TimeUnit.SECONDS)
+                .subscribe(emitter::onSuccess, error -> {
+                    if (attemptsLeft == 0 || ErrorInspector.contains(error, skipExceptions)) {
+                        emitter.onError(error);
+                    } else {
+                        call(single, emitter, jitteredDelaySec(attemptsLeft),
+                                attemptsLeft - 1, skipExceptions);
+                    }
+                });
+    }
+
+
+    /**
+     * Method returns jittered delay time in seconds.
+     * @param attemptsLeft number of attempts left.
+     * @return delay in seconds.
+     */
+    long jitteredDelaySec(int attemptsLeft) {
+        int numAttempt = maxAttempts - (maxAttempts - attemptsLeft);
+        double waitTimeSeconds =
+                Math.pow(2, ((numAttempt) % maxExponent))
+                        + jitterFactor * Math.random();
+        return (long) waitTimeSeconds;
+    }
```

Comment:
Have you thought about also adding max time?  (in addition to max tries)?  I can see cases that the total retry time can be unreasonably high and in that case we should stop retrying.

Reply:
Good idea. I added the max delay based on [Java script version of retry](https://github.com/aws-amplify/amplify-js/blob/df95ea3724eb6406f64b03f25086cd3e8644cb5f/packages/core/src/Util/Retry.ts#:~:text=function%20jitteredBackoff(maxDelayMs,args%2C%20jitteredBackoff(maxDelayMs))%3B) the max time is 300s. I feel that's too long. Any suggestion?

Reply:
Done.